### ISSUE TITLE
[Unified Language Service] Review REST API 

### ISSUE THREAD

**Issue by mikekistler** [Azure/azure-rest-api-specs#15686]
This API was reviewed prior to our current tracking process, so I am retroactively opening this issue to capture discussions on the API and any updates.

Service team rep is @rokulka .

Information from Rohan:

We are releasing this under the Unified Language Service scope, and all the APIs are still under preview – [Link](https://github.com/Azure/azure-rest-api-specs/tree/feature/cognitiveservices/language/specification/cognitiveservices/data-plane/Language/preview/2021-05-01-preview)

We went through a review earlier to discuss and decide the names and paths in the API - @Jeffrey Richter, @Johan Stenberg, @Mark Weitzel were part of it.

These discussions are noted in the below task items and some in the attached email.
[Feature 9795185: [Language] interface APIs for QnA and LUIS vNext](https://msazure.visualstudio.com/Cognitive%20Services/_workitems/edit/9795185)
[Feature 9795184: [Language] authoring/customization APIs - LUIS vNext](https://msazure.visualstudio.com/Cognitive%20Services/_workitems/edit/9795184)

**Comment by mikekistler** [comment]
It looks like there is a recording of the first review at

https://msit.microsoftstream.com/video/affa99ff-0300-a7ab-7079-f1ebae88c47c

But I do not have access.  Could someone please give me access?

**Comment by rokulka** [comment]
This review is for the Feedback API for the QuestionAnswering Service under the new Unified Language Service scenario.
 
We discussed the other granular update APIs in an earlier review, but the details of this API were not part of it. Hence requesting a review separately.
 
Description
In a Bot scenario, when a user queries our Knowledgebase – If the top QnAs in the response are have very close confidence scores, we provide a card suggesting ‘Do you mean:’ with questions from the top QnAs as options. When the User clicks one of the options, we provide the answer for that QnA. That’s where we call the Feedback API as well. This feedback information goes into our Active Learning scenario to generate question suggestions for QnAs.
 
API Details
```
POST /language/query-knowledgebases/projects/{projectName}/feedback
 
{
    "records": [
        {
            "UserId": "xyz",
            "UserQuestion": "what is notification panel",
            "QnaId": 2
        }
    ]
}

 
Response: No Content, Status Code 204
``` 
Notes
We don’t store the UserId, it goes into customer logs.
Request takes an array of records to provide flexibility to customers, allowing multiple feedback records in a single request.

We foresee only one operation for /feedback. This feedback is processed by our Active Learning algorithm to generate question suggestions to the QnAs. It will be made available as suggestions in the GET /qnas call -  [Link](https://github.com/Azure/azure-rest-api-specs/blob/ee3061aa89dcb77829a327cf159fbb9c12ea013f/specification/cognitiveservices/data-plane/Language/preview/2021-07-15-preview/questionanswering-authoring.json#L1171)

**Comment by mikekistler** [comment]
I think there was a misunderstanding during the API review so I’m hoping we can clear that up through a discussion in this issue, but if not we can schedule a follow-up meeting.

Verbs are allowed — actually recommended — in the “action” part of path, in the final path segment.  But verbs should not be used anywhere else in the path.  So for the Unified Language Service, the path 
"/:query-knowledgebases" is fine, since the verb “query” is part of the “action”.  But the path segment “/query-knowledebases” does not follow the Azure guidelines for URLs.

https://github.com/mkistler/api-guidelines/blob/lro/azure/Guidelines.md#uniform-resource-locators-urls

This path segment should be the name of a resource collection — a plural noun.  So to comply with our Guidelines, this path segment should be just “/knowledgebases”.

I can see how there might have been a misunderstanding here: “query-knowledgebases” is fine in one part of the path but not in another.  I suppose our Guidelines could also be more explicit about this.  But hope this has cleared up the confusion and there is no issue with making the changes need to comply with the guidelines.

**Comment by rokulka** [comment]
@mikekistler - Thanks for the recommendation. The Unified Language Service involves multiple cognitive services - Text Analytics, LUIS, QnA Maker and we went through multiple rounds of discussions to conclude on a set of path parameters for individual services. I will setup a sync to discuss this change in the path segment as it would affect all the services.

Please let me know if this /feedback API in particular looks fine other than the change in the path.

**Comment by mikekistler** [comment]
@rokulka I don't have any concerns beyond the verbs in the path segments.

**Comment by markweitzel** [comment]
# API REVIEW MEETING 1-Sept-21
Link to recording [here](https://microsoft.sharepoint.com/teams/AzureAPIStewardshipBoard/Shared%20Documents/General/Recordings/View%20Only/%5BUnified%20Language%20Service%5D%20Azure%20REST%20API%20Stewardship%20Board%20meeting%20for%20recommendations%20to%20API%20path-20210901_120213-Meeting%20Recording.mp4?web=1)
<sup>Internal Microsoft Link</sup>


This was revisiting the path structure of the Language Services, and specifically, the verb-noun (e.g. analyze-text), as well as the way that "actions" are denoted. In part, this comes down to authoring vs. inferencing. That is models are created (and ultimately deployed) through "projects" and then acted upon through actions. 

Below is the current plan of record. 
##Current Plan
> /language/analyze-text/projects/{projectName}
> /language/analyze-conversations/projects/{projectName}
> /language/query-knowledgebases/projects/{projectName}
> /language/:analyze-text
>/language/:analyze-conversations
> /language/:query-knowledgebases

There were several considerations discussed, e.g.
* The "overloading" of skill capability with an action, e.g. analyze-text (skill capability) and language/:analyze-text (an action to invoke)
* How do we create (and document) a pattern that can be expanded in a logical, prescriptive way

We also recognize that we need to maintain consistency w/in the entire cognitive services surface area as well. 

Next steps:
@rokulka will post the original meeting minutes (and possible recording) to this issue for review. 
@markweitzel & @mikekistler will review the original meeting minutes and, if necessary, propose modifications. 
**The deadline is Sept 8, 2021.**

**Comment by rokulka** [comment]
@markweitzel / @mikekistler  - Ted had pasted the details of the earlier discussion on the Teams chat. I'm pasting the doc link here for reference - [Link](https://microsoft.sharepoint-df.com/:w:/t/ALPSAzureLanguagePillars/ES26prT71QRHvwWVmZ_iRewBfMF1q5C2dHjqiF71S0-YlA?e=59bIgu)

**Comment by markweitzel** [comment]
@rokulka -- Thank you very much for providing the additional background information regarding the API structure. After reviewing this and meeting off-line with the folks on the stewardship board, we have a more complete understanding of the historical decisions. With that context, we explored various alternative strategies that would satisfy the requirements that have been articulated. None of the options we explored ended up being significantly better than what's already outlined. Given that, here is no reason to change course and we should continue down the path we are on. Thanks for giving us the opportunity to look at this again--all your help and support is greatly appreciated.
